server/src/test/java/org/elasticsearch/action/admin/indices/refresh/TransportUnpromotableShardRefreshActionTests.java

```diff
@@ -75,16 +113,197 @@ public void testRespondOKToRefreshRequestBeforeShardIsCreated() {
+                assertThat(future.isDone(), is(false));
+                assertThat(unpromotableShardOperationExecuted.get(), is(false));
+            }
+
+            setState(
+                clusterService,
+                ClusterState.builder(clusterService.state())
+                    .blocks(ClusterBlocks.builder().removeIndexBlock(shardId.getIndexName(), IndexMetadata.INDEX_REFRESH_BLOCK))
+            );
+        }
+
         assertThat(safeGet(future), sameInstance(ActionResponse.Empty.INSTANCE));
+        assertThat(unpromotableShardOperationExecuted.get(), is(true));
+    }
+
+    public void testActionWaitsUntilShardRefreshBlocksAreClearedMightTimeout() {
+        final var shardId = new ShardId(new Index(randomIdentifier(), randomUUID()), between(0, 3));
+        final var indexShardRoutingTable = createShardRoutingTableWithPrimaryAndSearchShards(shardId);
+
+        final IndicesService indicesService = mock(IndicesService.class);
+        // Register the action
+        new TransportUnpromotableShardRefreshAction(
+            clusterService,
+            transportService,
+            mock(ShardStateAction.class),
+            new ActionFilters(Set.of()),
+            indicesService,
+            threadPool,
+            true
+        ) {
+            @Override
+            protected void unpromotableShardOperation(
+                Task task,
+                UnpromotableShardRefreshRequest request,
+                ActionListener<ActionResponse.Empty> responseListener
+            ) {
+                assert false : "Unexpected call";
+            }
+        };
+
+        setState(
+            clusterService,
+            ClusterState.builder(clusterService.state())
+                .blocks(ClusterBlocks.builder().addIndexBlock(shardId.getIndexName(), IndexMetadata.INDEX_REFRESH_BLOCK))
+        );
+
+        final var countDownLatch = new CountDownLatch(1);
+        final var request = new UnpromotableShardRefreshRequest(
+            indexShardRoutingTable,
+            randomNonNegativeLong(),
+            randomNonNegativeLong(),
+            randomBoolean(),
+            TimeValue.timeValueSeconds(5)
+        );
+        transportService.sendRequest(localNode, TransportUnpromotableShardRefreshAction.NAME, request, expectError(e -> {
+            final Throwable rootCause = e.getRootCause();
+            assertThat(rootCause, instanceOf(ElasticsearchTimeoutException.class));
+            assertThat(rootCause.getMessage(), containsString("index refresh blocked, waiting for shard(s) to be started"));
+            countDownLatch.countDown();
+        }));
+
+        assertThat(countDownLatch.getCount(), is(equalTo(1L)));
+
+        if (randomBoolean()) {
+            setState(clusterService, ClusterState.builder(clusterService.state()).version(clusterService.state().version() + 1));
+            assertThat(countDownLatch.getCount(), is(equalTo(1L)));
```

Comment:
I wonder if we should remove this? If CI is super slow, we're at risk that the transport message times out before this instruction is executed?